src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp

```diff
@@ -45,294 +44,301 @@
-  _requested_generation(GLOBAL),
-  _degen_point(ShenandoahGC::_degenerated_outside_cycle),
-  _degen_generation(nullptr),
+  _requested_generation(nullptr),
+  _degen_point(ShenandoahGC::_degenerated_unset),
+  _heap(ShenandoahGenerationalHeap::heap()),
+  _age_period(0),
   _mode(none) {
   shenandoah_assert_generational();
   set_name("Shenandoah Control Thread");
   create_and_start();
 }
 
 void ShenandoahGenerationalControlThread::run_service() {
-  ShenandoahGenerationalHeap* const heap = ShenandoahGenerationalHeap::heap();
 
-  const GCMode default_mode = concurrent_normal;
-  ShenandoahGenerationType generation = GLOBAL;
-
-  uint age_period = 0;
+  const int64_t wait_ms = ShenandoahPacing ? ShenandoahControlIntervalMin : 0;
+  ShenandoahGCRequest request;
+  while (!should_terminate()) {
 
-  ShenandoahCollectorPolicy* const policy = heap->shenandoah_policy();
+    // This control loop iteration has seen this much allocation.
+    const size_t allocs_seen = reset_allocs_seen();
 
-  // Heuristics are notified of allocation failures here and other outcomes
-  // of the cycle. They're also used here to control whether the Nth consecutive
-  // degenerated cycle should be 'promoted' to a full cycle. The decision to
-  // trigger a cycle or not is evaluated on the regulator thread.
-  ShenandoahHeuristics* global_heuristics = heap->global_generation()->heuristics();
-  while (!in_graceful_shutdown() && !should_terminate()) {
     // Figure out if we have pending requests.
-    const bool alloc_failure_pending = _alloc_failure_gc.is_set();
-    const bool humongous_alloc_failure_pending = _humongous_alloc_failure_gc.is_set();
+    check_for_request(request);
 
-    GCCause::Cause cause = Atomic::xchg(&_requested_gc_cause, GCCause::_no_gc);
+    if (request.cause == GCCause::_shenandoah_stop_vm) {
+      break;
+    }
 
-    const bool is_gc_requested = ShenandoahCollectorPolicy::is_requested_gc(cause);
+    if (request.cause != GCCause::_no_gc) {
+      run_gc_cycle(request);
+    } else {
+      // Report to pacer that we have seen this many words allocated
+      if (ShenandoahPacing && (allocs_seen > 0)) {
+        _heap->pacer()->report_alloc(allocs_seen);
+      }
+    }
 
-    // This control loop iteration has seen this much allocation.
-    const size_t allocs_seen = reset_allocs_seen();
+    // If the cycle was cancelled, continue the next iteration to deal with it. Otherwise,
+    // if there was no other cycle requested, cleanup and wait for the next request.
+    if (!_heap->cancelled_gc()) {
+      MonitorLocker ml(&_control_lock, Mutex::_no_safepoint_check_flag);
+      if (_requested_gc_cause == GCCause::_no_gc) {
+        set_gc_mode(ml, none);
+        ml.wait(wait_ms);
+      }
+    }
+  }
 
-    // Check if we have seen a new target for soft max heap size.
-    const bool soft_max_changed = heap->check_soft_max_changed();
+  // In case any threads are waiting for a cycle to happen, let them know it isn't.
```

Comment:
maybe "it isn't happening", or "it won't happen".

Reply:
This is interesting. If GC is stopping prior to shutting down the VM, is there any point in notifying these waiting threads. Why not let them wait, and quietly shut things down? Are there JCK or other tests that would fail in that case?

Reply:
The waiting threads will remain waiting and prevent the JVM from shutting down.